src/main/java/com/bookbla/americano/domain/auth/service/impl/AuthServiceImpl.java

```diff
@@ -31,8 +35,40 @@ public LoginResponseDto login(LoginRequestDto loginRequestDto, String oAuthType)
         Member member = memberRepository.findByMemberTypeAndOauthEmail(memberType, oAuth2MemberResponse.getEmail())
             .orElseGet(() -> memberRepository.save(oAuth2MemberResponse.toMember()));
 
+        // 회원이 탈퇴 상태라면
+        if (member.getMemberStatus() == MemberStatus.DELETED) {
+            throw new BaseException(LoginExceptionType.CANNOT_LOGIN_MEMBER_DELETED);
+        }
+
         String accessToken = jwtProvider.createToken(member.getId().toString());
         return LoginResponseDto.of(accessToken, member);
     }
-  
+
+    @Override
+    public RejoinResponse rejoin(RejoinRequest rejoinRequest, String oAuthType) {
+        MemberType memberType = MemberType.from(oAuthType);
+        OAuth2Provider oAuth2Provider = oAuth2Providers.getProvider(memberType);
+        OAuth2MemberResponse oAuth2MemberResponse = oAuth2Provider.getMemberResponse(rejoinRequest.getAuthCode());
+
+        Member member = memberRepository.findByMemberTypeAndOauthEmail(memberType, oAuth2MemberResponse.getEmail())
+            .orElseGet(() -> memberRepository.save(oAuth2MemberResponse.toMember()));
+
+        LocalDateTime deleteAtThirtyDaysAfter = member.getDeleteAt().minusDays(30);
+
+        if (member.getMemberStatus() == MemberStatus.DELETED) {
+            // 회원 탈퇴 한 뒤에 30일이 지났다면
+            if (LocalDateTime.now().isAfter(deleteAtThirtyDaysAfter)) {
+                member.updateMemberStatus(MemberStatus.COMPLETED);
+            }
+            // 회원 탈퇴 한 뒤에 30일이 지나지 않았다면
+            else {
+               throw new BaseException(LoginExceptionType.NOT_THIRTY_DAYS_AFTER);
+            }
+        } else {
+            throw new BaseException(LoginExceptionType.NOT_MEMBER_DELETED);
+        }
```

Comment:
```suggestion
Member member = memberRepository.findByMemberTypeAndOauthEmailAndMemberStatusDeleted(memberType, oAuth2MemberResponse.getEmail(), MemberStatus.DELETED))
        .orElseThrow(() -> new BaseException(LoginExceptionType.NOT_MEMBER_DELETED);

LocalDateTime deleteAtThirtyDaysAfter = member.getDeleteAt().minusDays(30);

// 회원 탈퇴 한 뒤에 30일이 지났다면
if (LocalDateTime.now().isAfter(deleteAtThirtyDaysAfter)) {
    member.updateMemberStatus(MemberStatus.COMPLETED);
}
// 회원 탈퇴 한 뒤에 30일이 지나지 않았다면
else {
   throw new BaseException(LoginExceptionType.NOT_THIRTY_DAYS_AFTER);
}
```

재가입인 경우라, 새로 저장하진 않아도 될 것 같아서
DB에서 못찾아오는 경우에는 바로 예외를 던지는건 어떻게 생각하세요?

더불어 DB에서 삭제 회원 조건을 걸어 조회하는 건 어떻게 생각하시나용
아래 depth를 줄일 수 있을 것 같습니다

Reply:
슬 쿼리메서드가 길어지긴하네요!

Reply:
쿼리가 길어지니까 JPQL이나 QueryDSL로 바꾸긴해야겠네요 🤔